### PR TITLE
Optional pool freeing when using removeAll or retainAll

### PR DESCRIPTION
**Pull request by cypherdare**
These two functions could not be used if you want to free the removed items to a pool. The alternative would be to return the removed items, but that would create allocation/GC churn for something that is not always needed. And by far the most likely thing you would ever do with a removed item, if anything,  is return it to a pool.

I elected not to add this to `filter` as you would likely not use `filter` with pooled items. Pooled items are used in situations where the array is probably large and allocation/GC churn is trying to be avoided, so you would likely also want to be doing an in-place filter with `retainAll`.

### CODE REVIEW

File: collections/src/main/kotlin/ktx/collections/arrays.kt [hunk 1]
```diff
@@ -233,8 +235,9 @@ inline fun <Type, R : Comparable<R>> GdxArray<out Type>.sortByDescending(crossin
 
 /**
  * Removes elements from the array that satisfy the [predicate].
+ * @param pool Removed items are freed to this pool.
  */
-inline fun <Type> GdxArray<Type>.removeAll(predicate: (Type) -> Boolean) {
+inline fun <Type> GdxArray<Type>.removeAll(pool: Pool<Type>?, predicate: (Type) -> Boolean) {
```
Comment (czyzby) [inline collections/src/main/kotlin/ktx/collections/arrays.kt]: You should add a default `null` value for `pool` parameter.

File: collections/README.md [hunk 2]
```diff
@@ -32,7 +32,8 @@ chained.
 - Missing `addAll` and `removeAll` methods for arrays and iterables were added.
 - `iterate` method allows to iterate over collection's elements, while providing reference to `MutableInterator`. Can be
 used to easily remove collection elements during iteration.
-- `removeAll` and `retainAll` higher-order functions that work like collection extensions in Kotlin stdlib.
+- `removeAll` and `retainAll` higher-order functions that work like collection extensions  in Kotlin stdlib. A Pool can
```
Comment (czyzby) [inline collections/README.md]: Remove double space.

**Comment by czyzby** [comment]
Using `Pool` is a pretty specific use case. I think returning an array of the removed elements would be more flexible - although less performant and excessive in many cases. Consuming two lambdas, one of which would be used on the removed elements, is also an option, but then again - it makes the syntax more convoluted. Maybe we should add another method such as `selectAll` or something similar that would return the removed elements? I'm not sure.

**Comment by cypherdare** [comment]
Maybe this new method could take an optional array to put the removed
elements in? If null is passed, one could be allocated. This way there is
the option to use a reusable array. The function could return the other
array to allow immediate actions to be done on the removed elements.

**Comment by czyzby** [comment]
Also thought about this, since LibGDX seems to use this pattern sometimes. I think we should use this over the pool. Would you like to apply this change?

**Comment by cypherdare** [comment]
You mean add this function and remove the pool-freeing option?

And should it maybe be called something like “transfer” since “selectAll”
might sound ambiguous about whether the items are removed?

I would personally like an option to at least pass a removal handler lambda
to the other two methods if we’re removing the pool option.

**Comment by czyzby** [comment]
`transfer` sounds good, we can leave the `Pool` utilities and add `transfer` that consumes an `Array` instance and a lambda. Let me know if you want to add `transfer` in this PR or if you'd like to do this separately.

**Comment by cypherdare** [comment]
OK. I'll add it to this one. Could be a few days.

**Comment by czyzby** [comment]
No problem, thanks again.

**Comment by czyzby** [comment]
I'll go ahead and finish the PR, I hope you didn't start working on it yet. Thanks for the contribution!

**Comment by czyzby** [comment]
Done, you can view the changes in the latest linked commit.

**Pull request merged**